src/compartment.c

```diff
@@ -958,59 +981,80 @@ resolve_rela_syms(struct Compartment *new_comp)
                 continue;
             }
 
-            // Prioritise looking for weak symbols in libraries outside the
-            // source library, even if they are defined
-            if (curr_rela_map->rela_sym_bind == STB_WEAK)
+            candidate_syms = comp_syms_find_all(
+                curr_rela_map->rela_name, new_comp->comp_syms);
+
+            if (candidate_syms->data_count == 0)
             {
-                found_sym = find_lib_dep_sym_in_comp(curr_rela_map->rela_name,
-                    new_comp, i, curr_rela_map->rela_sym_type, false);
-                if (!found_sym.found)
+                if (curr_rela_map->rela_sym_bind == STB_WEAK)
                 {
-                    found_sym
-                        = find_lib_dep_sym_in_comp(curr_rela_map->rela_name,
-                            new_comp, i, curr_rela_map->rela_sym_type, true);
+                    // TODO Hack to suppress weak `libc` relocations
+                    if (strcmp(new_comp->libs[i]->lib_name, "libc.so.7"))
```

Comment:
Should we make this a bit more abstract e.g. at least "if name starts with libc.so" (i.e. without the ".7")?

Reply:
I would've liked to expand this into providing a list of libraries to suppress, or parameterise it, but maybe that's too much at this point, but at least I did the prefix check for `libc.so` in https://github.com/capablevms/CHERI-ELF-comp/pull/38/commits/be0d58b12eef0a43cf1e906d96177eb3747446fa.